### ISSUE TITLE
FSDP CPU Offload + use_orig_params + Prefetch Race Condition Leading to Loss Mismatch on Slow Compute Hardware

### ISSUE THREAD

**Issue by aeeeeeep** [pytorch/pytorch#178142]
### 🐛 Describe the bug

When using FSDP with `cpu_offload=True`, `sharding_strategy=NO_SHARD`, and `use_orig_params=True`, combined with `forward_prefetch=True` or `backward_prefetch=BackwardPrefetch.BACKWARD_POST`, there is a numerical mismatch in the loss values starting from the second training step.

This issue **only reproduces on specific slow-compute hardware** and **cannot be reproduced on NVIDIA GPUs**. Due to NDA restrictions, we cannot disclose the specific hardware details. The root cause is suspected to be a race condition in stream synchronization during prefetch operations, which becomes more pronounced on hardware with slower GPU compute speeds relative to CPU.

### Reproduction Code

```python3
import os
import torch
import torch.distributed as dist
import torch.nn as nn
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP, CPUOffload
from torch.distributed.fsdp import ShardingStrategy, BackwardPrefetch
from dataclasses import dataclass
from typing import Optional


@dataclass
class Config:
    num_layers: int = 2
    use_orig_params: bool = True
    forward_prefetch: bool = False
    backward_prefetch: Optional[BackwardPrefetch] = None
    batch_size: int = 1
    input_dim: int = 1
    hidden_dim: int = 1

def build_base_model(config):
    dims = [config.input_dim] + [config.hidden_dim] * (config.num_layers - 1) + [config.input_dim]
    layers = [nn.Linear(dims[i], dims[i+1]) for i in range(config.num_layers)]
    
    return nn.Sequential(*layers)

def wrap_fsdp(module, config):
    return FSDP(
        module,
        cpu_offload=CPUOffload(offload_params=True),
        sharding_strategy=ShardingStrategy.NO_SHARD,
        use_orig_params=config.use_orig_params,
        forward_prefetch=config.forward_prefetch,
        backward_prefetch=config.backward_prefetch,
    )

def run_training(rank, config, use_fsdp):
    torch.manual_seed(0)
    model = build_base_model(config).cuda(rank)
    
    if use_fsdp:
        for name, child in model.named_children():
            setattr(model, name, wrap_fsdp(child, config))
        model = wrap_fsdp(model, config)

    losses = []

    for step in range(2):
        torch.manual_seed(step + 1000)
        x = torch.rand(config.batch_size, config.input_dim, device='cuda')
        loss = model(x).sum()
        # !!! fix
        # torch.cuda.synchronize()
        # !!! fix
        loss.backward()
        losses.append(loss.detach().cpu())

    return losses

def test_config(rank, config):
    ddp_losses = run_training(rank, config, use_fsdp=False)
    fsdp_losses = run_training(rank, config, use_fsdp=True)

    if rank == 0:
        try:
            for dl, fl in zip(ddp_losses, fsdp_losses):
                torch.testing.assert_close(dl, fl, rtol=1e-4, atol=1e-4)
            print(f"✅ Pass: prefetch(fwd={config.forward_prefetch}, bwd={config.backward_prefetch})")
        except AssertionError as e:
            print(f"❌ Fail: prefetch(fwd={config.forward_prefetch}, bwd={config.backward_prefetch})")

def setup_dist(rank, world_size):
    os.environ.update({
        "MASTER_ADDR": "127.0.0.1",
        "MASTER_PORT": "36666",
    })
    torch.cuda.set_device(rank)
    dist.init_process_group('nccl', rank=rank, world_size=world_size)

if __name__ == "__main__":
    configs = [
        Config(),
        Config(forward_prefetch=True),
        Config(backward_prefetch=BackwardPrefetch.BACKWARD_PRE),
        Config(backward_prefetch=BackwardPrefetch.BACKWARD_POST),
        Config(forward_prefetch=True, backward_prefetch=BackwardPrefetch.BACKWARD_PRE),
        Config(forward_prefetch=True, backward_prefetch=BackwardPrefetch.BACKWARD_POST),
    ]

    setup_dist(0, 1)
    for cfg in configs:
        test_config(0, cfg)
```

### Actual Output (NVIDIA GPU all pass)
```
✅ Pass: prefetch(fwd=False, bwd=None)
❌ Fail: prefetch(fwd=True, bwd=None)
✅ Pass: prefetch(fwd=False, bwd=BackwardPrefetch.BACKWARD_PRE)
❌ Fail: prefetch(fwd=False, bwd=BackwardPrefetch.BACKWARD_POST)
✅ Pass: prefetch(fwd=True, bwd=BackwardPrefetch.BACKWARD_PRE)
❌ Fail: prefetch(fwd=True, bwd=BackwardPrefetch.BACKWARD_POST)
```

### Workaround Fix

The following workaround adds stream synchronization before `pre_unshard()` when both `cpu_offload` and `use_orig_params` are enabled. This ensures that `_writeback_orig_params()` reads consistent parameter states.

```python
@no_type_check
def _unshard(
    state: _FSDPState,
    handle: FlatParamHandle,
    unshard_stream: torch.Stream,
    pre_unshard_stream: torch.Stream,
) -> None:
    """
    Unshards the handles in ``handles``. If the handles are in
    :meth:`summon_full_params` and are using mixed precision, then they are
    forced to full precision.

    Postcondition: handle's ``FlatParameter`` 's data is the padded
    unsharded flat parameter on the compute device.
    """
    if not handle:
        return
    
    # Workaround: Ensure pre_unshard_stream waits for current stream when
    # cpu_offload and use_orig_params are both enabled. This prevents race
    # conditions in _writeback_orig_params() during prefetch operations.
    if handle._offload_params and handle._use_orig_params:
        pre_unshard_stream.wait_stream(state._device_handle.current_stream())
    
  with state._device_handle.stream(pre_unshard_stream):
        ran_pre_unshard = handle.pre_unshard()
    if ran_pre_unshard:
        unshard_stream.wait_stream(pre_unshard_stream)
    if state.limit_all_gathers:
        event = state._free_event_queue.dequeue_if_needed()
        if event:
            with torch.profiler.record_function(
                "FullyShardedDataParallel.rate_limiter"
            ):
                event.synchronize()
    with state._device_handle.stream(unshard_stream):
        handle.unshard()
        handle.post_unshard()
```

### Maybe Root Cause

When forward_prefetch=True or backward_prefetch=BACKWARD_POST, _prefetch_handle() calls _unshard() asynchronously without synchronization. The pre_unshard() function executes _writeback_orig_params() on _pre_unshard_stream, which reads parameter states that may still be modified by the current computation stream.

On slow GPU hardware, the timing difference between async operations becomes more pronounced, making this race condition more likely to occur.

### Notes

- This fix is a **workaround** and may not address the root cause completely
- Cannot reproduce on NVIDIA GPUs

### Versions

Version details cannot be disclosed due to NDA

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @xmfan @zhaojuanmao @mrshenli @rohan-varma @chauhang @mori360 @ppwwyyxx
